File: CMSI386-ProgrammingLanguages/homework2/warmup.py
import random
import math
import requests
from Crypto.Cipher import AES

def change(amount):
    if amount < 0:
        raise ValueError('amount cannot be negative')
    remaining = amount
    result = [None] * 4
    for index, value in enumerate((25, 10, 5, 1)):
        result[index], remaining = divmod(remaining, value)
    return tuple(result)

def strip_quotes(s):
    return s.translate(str.maketrans('', '', '\'"'))

def scramble(s):
    return ''.join(random.sample(s, len(s)))

def say(first_word=None):
    words = []
    def say_more(word=None):
        if word is None:
            return ' '.join(words)
        words.append(word)
        return say_more
    return say_more(first_word)

def triples(limit):
    result = []
    for c in range(1, limit + 1):
        for b in range(1, c):
            for a in range(1, b):
                if a * a + b * b == c * c:
                    result.append((a, b, c))
    return sorted(result, key=lambda x: x[0])

def powers(base, limit):
    current_value = 1
    while current_value <= limit:
        yield current_value
        current_value *= base

def interleave(a, *b):
    first_list, second_list, result = a, list(b), []
    while bool(first_list) and bool(second_list):
        result.append(first_list.pop(0))
        result.append(second_list.pop(0))
    return result + first_list + second_list

class Cylinder():
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height
    @property
    def volume(self):
        return math.pi * (self.radius ** 2) * self.height
    @property
    def surface_area(self):
        return (2 * math.pi * (self.radius * self.height)) + (2 * math.pi * (self.radius ** 2))
    def widen(self, factor):
        self.radius *= factor
    def stretch(self, factor):
        self.height *= factor

def make_crypto_functions(key, iv):
    def encrypt(msg):
        return AES.new(key, AES.MODE_CBC, iv).encrypt(msg)
    def decrypt(cypher):
        return AES.new(key, AES.MODE_CBC, iv).decrypt(cypher)
    return [encrypt, decrypt]

def random_name(**payload):
    payload.update({'amount':1})
    r = requests.get('https://uinames.com/api/', params=payload)
    res = r.json()
    if 'error' in res:
        raise ValueError(res)
    return "{}, {}".format(res['surname'], res['name'])

File: CMSI386-ProgrammingLanguages/homework1/src/warmup.js
const crypto = require('crypto'); // eslint-disable-line import/no-extraneous-dependencies
const rp = require('request-promise');

const change = (cents) => {
  const numberOfCoins = [];
  let centsRemaining = cents;
  if (centsRemaining < 0) { throw new RangeError('amount cannot be negative'); }
  [25, 10, 5, 1].forEach((element) => {
    numberOfCoins.push(Math.floor(centsRemaining / element));
    centsRemaining %= element;
  });
  return numberOfCoins;
};

const stripQuotes = (string) => {
  const stringToStrip = string;
  return stringToStrip.split('').filter(char => char !== '"' && char !== '\'').join('');
};

const scramble = (string) => {
  const stringArray = string.split('');
  let finalString = [];
  // iterates over copy of stringArray, safe from mutation
  stringArray.slice().forEach((val, index) => {
    const randomNumber = Math.floor(Math.random() * (string.length - index));
    const randomChar = stringArray.splice(randomNumber, 1);
    finalString = finalString.concat(randomChar);
  });
  return finalString.join('');
};

const powers = (base, limit, callback) => {
  let currentValue = 1;
  if (currentValue < 1) { return; }
  while (currentValue <= limit) {
    callback(currentValue);
    currentValue *= base;
  }
};

// Arrow functions cannot be used as generator functions
function* powersGenerator(base, limit) {
  let currentValue = 0;
  while (base ** currentValue <= limit) {
    yield base ** currentValue;
    currentValue += 1;
  }
}

const say = (msg) => {
  const out = [];
  function sayCurrier(nextMsg) {
    if (!nextMsg) {
      return out.join(' ');
    }
    out.push(nextMsg);
    return sayCurrier;
  }
  return sayCurrier(msg);
};

const interleave = (arrayA, ...arrayB) => {
  const arrA = arrayA;
  const arrB = arrayB;
  const interleavedArr = [];

  while (arrA.length > 0 || arrB.length > 0) {
    if (arrA.length > 0) { interleavedArr.push(arrA.shift()); }
    if (arrB.length > 0) { interleavedArr.push(arrB.shift()); }
  }
  return interleavedArr;
};

const cylinder = (spec) => {
  let { radius = 1, height = 1 } = spec;
  const volume = () => Math.PI * (radius ** 2) * height;
  const surfaceArea = () => (2 * Math.PI * (radius * height)) + (2 * Math.PI * (radius ** 2));
  const widen = (factor) => { radius *= factor; };
  const stretch = (factor) => { height *= factor; };
  const toString = () => `Cylinder with radius ${radius} and height ${height}`;
  return Object.freeze({
    volume,
    surfaceArea,
    widen,
    stretch,
    toString,
    get radius() { return radius; },
    get height() { return height; },
  });
};

const makeCryptoFunctions = (cryptoKey, cryptoAlg) => {
  // Adapted from: http://lollyrock.com/articles/nodejs-encryption/
  const encrypt = (text) => {
    const cipher = crypto.createCipher(cryptoAlg, cryptoKey);
    let crypted = cipher.update(text, 'utf8', 'hex');
    crypted += cipher.final('hex');
    return crypted;
  };

  const decrypt = (text) => {
    const decipher = crypto.createDecipher(cryptoAlg, cryptoKey);
    let dec = decipher.update(text, 'hex', 'utf8');
    dec += decipher.final('utf8');
    return dec;
  };
  return [encrypt, decrypt];
};

const randomName = (params) => {
  const { gender, region } = params;
  return rp({
    method: 'GET',
    uri: 'https://uinames.com/api/',
    qs: {
      gender,
      region,
      amount: 1,
    },
    headers: {
      'User-Agent': 'LMU Programming Languages HW',
    },
    json: true, // Automatically parses the JSON string in the response
  }).then(data => `${data.surname}, ${data.name}`);
};

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  interleave,
  cylinder,
  makeCryptoFunctions,
  randomName,
};
